File: 2021/Day_07/code.php
<?php

require_once __DIR__ . '/../../lib/php/utils.php';

$input = readInput(__DIR__, ',');

// Task code
function part01(array $input)
{
  return min(array_map(function ($v) use ($input){
    return array_sum(array_map(function ($c_r_a_b_o_i_d_b_e_r_g) use ($v) {
      return abs($c_r_a_b_o_i_d_b_e_r_g - $v);
    }, $input));
  }, range(min($input), max($input)))
  );
}

function part02(array $input)
{
  return min(array_map(function ($v) use ($input){
    return array_sum(array_map(function ($c_r_a_b_o_i_d_b_e_r_g) use ($v) {
      return (abs($c_r_a_b_o_i_d_b_e_r_g - $v) * (abs($c_r_a_b_o_i_d_b_e_r_g - $v )+1)) / 2;
    }, $input));
  }, range(min($input), max($input)))
  );
}

// Execute
calcExecutionTime();
$result01 = part01($input);
$result02 = part02($input);
$executionTime = calcExecutionTime();

writeln('Solution Part 1: ' . $result01);
writeln('Solution Part 2: ' . $result02);
writeln('Execution time: ' . $executionTime);

saveBenchmarkTime($executionTime, __DIR__);

// Task test
testResults(
    [], // Expected
    [$result01, $result02], // Result
);

File: 2021/Day_01/code.go
package main

import (
	"advent-of-code/helper"
	"fmt"
	"path"
	"runtime"
	"time"
)

func main() {
	_, f, _, _ := runtime.Caller(0)
	cwd := path.Join(path.Dir(f))

	input := helper.ReadInput(cwd, helper.NewLine)
	iValues := input.Ints()

	// Execute
	start := time.Now()
	result01 := part01(iValues)
	result02 := part02(iValues)
	executionTime := helper.ExecutionTime(time.Since(start))

	fmt.Printf("Solution Part 1: %v\n", result01)
	fmt.Printf("Solution Part 2: %v\n", result02)
	fmt.Printf("Execution time: %s\n", executionTime)

	helper.SaveBenchmarkTime(executionTime, cwd)

	// Testing
	helper.TestResults(
		[]helper.TestingValue{
			helper.TestingValue{Result: result01, Expect: 1502},
			helper.TestingValue{Result: result02, Expect: 1538},
		},
	)
}

// Task code
func part01(input []int) (count int) {
  var prev int
  for i, value := range input {
    if i == 0 {
      prev = value
    } else {
      if value > prev {
        count ++
      }
      prev = value
    }
  }
  return 
}

func part02(input []int) (count int) {
  for i, value := range input {
    if i == 0 || i == 1 || i == 2 {
      continue
    }
    if value > input[i-3] {
      count ++
    }
  }
  return
}
